src/Pool.sol

```diff
@@ -141,157 +157,243 @@ abstract contract Pool is Pausable, Ownable, ERC4626 {
+        delete _withdrawFees[withdrawFeeId_]; // TO-DO Check if leaving a gap could be avoidable or doesnt matter
     }
 
     /**
      * @dev Adds a new deposit.
-     * @param amount.
+     * @param amount_ Amount to be deposited.
      */
-    function deposit(uint32 amount) external whenNotPaused {
-        /*
-            1. Adds the deposit.
-            2. Charges the deposit fee into the recipient address.
-        */
+    function deposit(uint128 amount_) external whenNotPaused nonReentrant {
+        // Update the rewards per token.
+        updateRewardsPerToken();
+
+        // Get the total deposit taking into account the deposit fee.
+        uint32 totalFee_ = uint32(_depositFee * MULTIPLIER / 100);
+        uint32 depositedAmount_ = uint32(amount_ - ((amount_ * totalFee_) / MULTIPLIER));
+        if (_depositAsset.balanceOf(msg.sender) < depositedAmount_) revert Pool_NotEnoughBalanceToDepositError();
+
+        // Transfer the deposit fee into the recipient address.
+        bool success_ = _depositAsset.transferFrom(msg.sender, _depositFeeRecipient, amount_ - depositedAmount_);
+        if (!success_) revert Pool_DepositFeeChargeFailedError();
+
+        // Add the deposit.
+        _usersData[msg.sender].deposits.push(DepositData(uint64(block.number), depositedAmount_));
+
+        // Transfer the assets into the Pool.
+        success_ = _depositAsset.transferFrom(msg.sender, address(this), depositedAmount_);
+        if (!success_) revert Pool_AssetTransferFailedError();
+
+        // Update the totalAmountDeposited by the User.
+        _usersData[msg.sender].totalAmountDeposited += depositedAmount_;
+
+        // Update the total deposited amount.
+        _totalAmount += depositedAmount_;
+
+        // Update the next Deposit to be removed if it's the first Deposit of the User.
+        if (_usersData[msg.sender].deposits.length == 1) {
+            _usersData[msg.sender].nextDepositIdToRemove = 0;
+        }
     }
 
     /**
      * @dev Withdraw a deposit.
-     * @param depositId ID of the deposit.
+     * @param amount_ Amount to be withdrawed.
      */
-    function withdraw(uint32 depositId) external {
-        /*
-            1. Checks if the deposit exists.
-            2. Withdraws the deposit (Taking into account the different deposits)
-        */
+    function withdraw(uint128 amount_) external nonReentrant {
+        // Update the rewards per token.
+        updateRewardsPerToken();
+
+        // Check if the user has deposited enough amount as the requested to withdraw.
+        if (_usersData[msg.sender].totalAmountDeposited < amount_) revert Pool_NotEnoughAmountDepositedError();
+
+        // Check user's deposits in order to compute the amount to withdraw.
+        uint128 withdrawAmount_;
+        uint128 totalFee_;
+        for (uint32 i = _usersData[msg.sender].nextDepositIdToRemove; i < _usersData[msg.sender].deposits.length; i++) {
```

Comment:
info: this for loop can be changed to a while loop with the condition `(withdrawAmount_ >= amount_)`.